### ISSUE TITLE
Cannot proxy kube-apiserver

### ISSUE THREAD

**Issue by joejulian** [containous/traefik#6527]
### Do you want to request a *feature* or report a *bug*?

Bug

### What did you do?

```
cd $(mktemp -d)
export KUBECONFIG=$PWD/conf
kind create cluster

# download values and ingress
curl -L https://gist.github.com/joejulian/ebba694a625333dc989a288a221c04cd/archive/0eecf28355c1a11d1965457a0cf44202ca972ded.tar.gz | tar xzv --strip 1

# deploy traefik and create ingress 
helm install traefik stable/traefik --values traefik_values.yaml
kubectl apply -f ingress.yaml

# create token account for testing. This represents an analog for oidc auth.
kubectl create sa test
kubectl get clusterrolebinding cluster-admin -o yaml | yq '.subjects[.subjects | length] |=.+ {kind: "ServiceAccount", name: "test", namespace: "default"}' | kubectl apply -f -

# set some variables to use when we switch configs
TOKEN="$(kubectl get secret "$(kubectl get serviceaccounts test -o jsonpath='{.secrets[0].name}')" -o jsonpath='{.data.token}' | base64 -d)"
POD="$(kubectl get pod -l app=traefik -o jsonpath='{.items[0].metadata.name}' )"

# port forward to traefik since we can't reach the node port from the host
kubectl port-forward $POD 8443:443 &

# create a kubeconfig that uses token auth
yq -y '.users[0]={name: "kind-kind", user: { token: "'$TOKEN'" }}' conf | yq -y '.clusters[0].cluster.server="https://127.0.0.1:8443/testpath"' > conf-token

kubectl --kubeconfig conf-token --insecure-skip-tls-verify -v9 port-forward $POD 8080
```
### What did you expect to see?
```
Forwarding from 127.0.0.1:8080 -> 8080
Forwarding from [::1]:8080 -> 8080
```

### What did you see instead?
```
Handling connection for 8443
Handling connection for 8443
error: error upgrading connection: unable to upgrade connection: Internal Server Error
```

### Output of `traefik version`: (_What version of Traefik are you using?_)

```
Version:      v1.7.20
Codename:     maroilles
Go version:   go1.12.14
Built:        2019-12-10_05:02:10PM
OS/Arch:      linux/amd64
```

### What is your environment & configuration (arguments, toml, provider, platform, ...)?

```toml
# set by chart template
```

### If applicable, please paste the log output in DEBUG level (`--log.level=DEBUG` switch)

_filtered out 'event kind'_
https://gist.github.com/joejulian/dd2ae3cdb1de5d9354923fab6e1a69a1

**Comment by joejulian** [comment]
This happens because the kube-apiserver can only create streams using the deprecated SPDY protocol and has [not been able to reach a consensus](https://github.com/kubernetes/kubernetes/issues/7452) on how to move forward with supporting http2.

We would like to help with adding http2 support to k/k but that's a much longer-term project. We have users, today, that need a much nearer term fix. To that end, we're working on a set of PRs to solve this bug.

**Comment by ldez** [comment]
https://community.containo.us/t/ingress-doesnt-work-with-connection-upgrade/889/11?u=ldez

**Comment by joejulian** [comment]
Try the repro. I'm not using curl and I haven't and couldn't make a header typo. It works without traefik so I think that statement is wrong.

**Comment by joejulian** [comment]
I've also tested sometime around last December with v2 with the same results.

**Comment by joejulian** [comment]
adding -v7 to the kubectl command:
```
...
I0319 13:44:58.204917  287601 round_trippers.go:420] POST https://127.0.0.1:8443/testpath/api/v1/namespaces/default/pods/traefik-7948558cf5-bx7m8/portforward
I0319 13:44:58.204934  287601 round_trippers.go:427] Request Headers:
I0319 13:44:58.204946  287601 round_trippers.go:431]     X-Stream-Protocol-Version: portforward.k8s.io
I0319 13:44:58.204955  287601 round_trippers.go:431]     User-Agent: kubectl/v1.17.4 (linux/amd64) kubernetes/8d8aa39
I0319 13:44:58.204965  287601 round_trippers.go:431]     Authorization: Bearer <masked>
I0319 13:44:58.214811  287601 round_trippers.go:446] Response Status: 500 Internal Server Error in 9 milliseconds
F0319 13:44:58.214970  287601 helpers.go:114] error: error upgrading connection: unable to upgrade connection: Internal Server Error
```
which corresponds with 
```
default/traefik-7948558cf5-bx7m8[traefik]: {"level":"debug","msg":"'500 Internal Server Error' caused by: http2: invalid Upgrade request header: [\"spdy/3.1\"]","time":"2020-03-19T20:53:54Z"}
```
probably because http/2 can't be upgraded to SPDY. Similar to how it works with websocket, I believe, the connection will need to be reestablished with http/1.1 and then upgraded.

**Comment by joejulian** [comment]
```
8.1.2.2.  Connection-Specific Header Fields

...

      Note: HTTP/2 purposefully does not support upgrade to another
      protocol.  The handshake methods described in Section 3 are
      believed sufficient to negotiate the use of alternative protocols.
```

**Comment by juliens** [comment]
Hi @joejulian , 
as explain on the forum, for me it works because `kubectl` (in the TLS-ALPN) force the use of http1.1 (whereas curl doesn't), and the kubernetes api server force to use http1.1 too. 

So may be it's a problem with your `kubectl version`. Could you provide us your kubectl version?

Or it can be the kubernetes api implementation. I tried with `k3d` and not with `kind`, I will try with `kind` as soon as possible.

**Comment by joejulian** [comment]
```
Client Version: version.Info{Major:"1", Minor:"17", GitVersion:"v1.17.4", GitCommit:"8d8aa39598534325ad77120c120a22b3a990b5ea", GitTreeState:"clean", BuildDate:"2020-03-12T21:03:42Z", GoVersion:"go1.13.8", Compiler:"gc", Platform:"linux/amd64"}
Server Version: version.Info{Major:"1", Minor:"17", GitVersion:"v1.17.0", GitCommit:"70132b0f130acc0bed193d9ba59dd186f0e634cf", GitTreeState:"clean", BuildDate:"2020-01-14T00:09:19Z", GoVersion:"go1.13.4", Compiler:"gc", Platform:"linux/amd64"}
```

**Comment by traefiker** [comment]
Closed by #6553.

**Comment by traefiker** [comment]
Closed by #6554.

**Comment by joejulian** [comment]
I have confirmed that this does solve the issue. Thank you.
